src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPDocumentSynchronizer.cs

```diff
@@ -94,45 +135,77 @@ internal void DocumentManager_Changed(object sender, LSPDocumentChangeEventArgs
+                    {
+                        var virtualDocument = lspDocument.VirtualDocuments[i];
 
-                if (lspDocument.Version == synchronizingContext.ExpectedHostDocumentVersion)
-                {
-                    synchronizingContext.SetSynchronized(true);
+                        Debug.Assert(!_virtualDocumentContexts.ContainsKey(virtualDocument.Uri));
+
+                        var virtualDocumentTextBuffer = virtualDocument.Snapshot.TextBuffer;
+                        virtualDocumentTextBuffer.PostChanged += VirtualDocumentBuffer_PostChanged;
+                        _virtualDocumentContexts[virtualDocument.Uri] = new DocumentContext(_synchronizationTimeout);
+                    }
                 }
-                else if (lspDocument.Version > synchronizingContext.ExpectedHostDocumentVersion)
+                else if (args.Kind == LSPDocumentChangeKind.Removed)
                 {
-                    // The LSP document version has surpassed what the projected document was expecting for a version. No longer able to synchronize.
-                    synchronizingContext.SetSynchronized(false);
+                    var lspDocument = args.Old;
+                    for (var i = 0; i < lspDocument.VirtualDocuments.Count; i++)
+                    {
+                        var virtualDocument = lspDocument.VirtualDocuments[i];
+
+                        Debug.Assert(_virtualDocumentContexts.ContainsKey(virtualDocument.Uri));
+
+                        var virtualDocumentTextBuffer = virtualDocument.Snapshot.TextBuffer;
+                        virtualDocumentTextBuffer.PostChanged -= VirtualDocumentBuffer_PostChanged;
+                        _virtualDocumentContexts.Remove(virtualDocument.Uri);
+                    }
                 }
             }
         }
 
-        private class DocumentSynchronizingContext
+        private class DocumentContext
+        {
+            private readonly TimeSpan _synchronizingTimeout;
+
+            public DocumentContext(TimeSpan synchronizingTimeout)
+            {
+                _synchronizingTimeout = synchronizingTimeout;
+            }
+
+            public long SeenHostDocumentVersion { get; private set; }
+
+            public DocumentSynchronizingContext SynchronizingContext { get; private set; }
+
+            public void UpdateSeenDocumentVersion(long seenDocumentVersion)
+            {
+                SeenHostDocumentVersion = seenDocumentVersion;
+            }
+
+            public void UpdateSynchronizingContext(int requiredHostDocumentVersion, CancellationToken cancellationToken)
+            {
+                // Cancel our existing synchronizing context.
+                SynchronizingContext?.SetSynchronized(result: false);
+                SynchronizingContext = new DocumentSynchronizingContext(requiredHostDocumentVersion, _synchronizingTimeout, cancellationToken);
+            }
+        }
+
+        public class DocumentSynchronizingContext
```

Comment:
A public class inside of an internal class seems weird, does this need to be private or in its own file?

Reply:
Should make it private 😄